cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java

```diff
@@ -98,6 +104,46 @@ public static void validateConfig(SecureStore secureStore, SourceControlConfig s
     }
   }
 
+  /**
+   * Returns the <a href="https://git-scm.com/docs/git-hash-object">Git Hash</a>
+   * of the requested file path in the current HEAD. For symlinks, it returns the hash of
+   * the target file. This ensures that the file hash of a symlink is equal to the hash of the target file. It
+   * doesn't return the hash for directories. If the directory contains symlinks, we can't depend on the
+   * directory hash to ensure the contents are unchanged.
+   *
+   * @param path The absolute path on the filesystem that should include {@link RepositoryManager#getBasePath()} as a
+   *             prefix.
+   * @return The git file hash of the requested file.
+   * @throws IOException              when file isn't found, there are circular symlinks or other file IO errors.
+   * @throws IllegalArgumentException when the provided file isn't a regular file, eg: directory.
+   * @throws NotFoundException        when the file isn't committed to Git.
+   * @throws IllegalStateException    when {@link RepositoryManager#startUp()} isn't called before this.
+   */
+  public String getFileHash(Path path) throws IOException, NotFoundException {
```

Comment:
I'm not really clear on how we'll be using this method, but it seems incorrect to have this.

I'm guessing this is for a pull pipeline API call, where we want to deploy a specific pipeline? If that's the case, there needs to be one method that both returns the content of the file as well as the hash, not a separate method that only returns the hash. Otherwise there will be race conditions where we get the file content, HEAD changes, and we fetch a file hash that does not match the content.